### PR TITLE
Implement Toolbar in Redux

### LINKED ISSUE #1576: Implement toolbar in redux
**Implement toolbar in react/redux**
- [x] image download
- [x] Projection selection
- [x] Info list
- [x] Feedback module
- [x] Notifications
- [x] Share feature

**Necessary Updates**
Put together a framework for incrementally migrating to `Redux`
- [x] Create reusable react modal component
- [x] Create reusable react list component
- [x] Implement standard (at least semi-standard) Redux File Structure
- [x] Implement crutch between old state system and new redux
- [x] Find `React` replacement for `Jcrop`
- [x] Make component that updates permalink 
- [x] Add documentation
- [x] Update tests

### PR DESCRIPTION
**Pull request by Benjaki2**
## Description

Fixes #1576 

- [x] Syncs redux state with old models
- [x] Create reusable redux/react modal component
- [x] Create reusable redux/react list component
- [x] Implement standard (at least semi-standard) Redux File Structure
- [x] Maintain permalink with spotify/redux-location-state
- [x] Add documentation
- [x] Update tests
- [x] Create reusable redux/react bounding-box selection component for image download and GIFs (not impletemented in GIF yet)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/nasa-gibs/worldview/blob/master/.github/CONTRIBUTING.md) doc
- [x] I have added necessary documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [x] Any dependent changes have been merged and published in downstream modules (if applicable)

**Bugs found**
- [x] Uncaught TypeError: Cannot read property 'length' of null
    at Object.transform (proj.js:467)
    at ImageDownloadContainer.render
- [x] Styling of about page
- [ ] Make permalink listener only listens for permalink changes when permalink is open
- [x] X to close modals style is difficult to see in some dialogs/modals
- [x] Layer settings scroll bar is black and hard to see
- [x] The 'Story Complete' dialog background change from light to dark makes the text hard to see
- [x] (IE11 only?) The 'Explore Worldview' line breaks so that 'Worldview' isn't on the same line
- [x] When changing projection, the toolbar remains open after changing to the selected projection
- [x] Double clicking the Info Toolbar triggers feedback to be invoked
- [ ] Image download button onClick toggle action instead of Open

## Permalink handling
Permalink handling is initiated in the `./js/location.js` file. This looks pretty funky right now. As legacy models are replaced by the redux state this should become really clean. The projection model is the only models that has been replaced in the PR. All that is needed to retain the projection state properly in the permalink is [this object](https://github.com/nasa-gibs/worldview/blob/redux-toolbar-1576/web/js/location.js#L26-L31):
```
  p: {
    stateKey: 'proj.id',
    initialState: 'geographic'
  }
```

## File Setup:

**Components**
Pretty much the same

**Modules** 
The module folder contains `actions`, `reducers`, `constants`, `tests`, `selectors`(none yet) and `util` files associated with any one module

**Containers**
This is generally where an outer template of an individual component is located. These templates retrieve the state via the redux [connect](https://react-redux.js.org/api/connect) method and pass those as props down to child components. [See Example](https://github.com/nasa-gibs/worldview/blob/redux-toolbar-1576/web/js/containers/toolbar.js#L118-L136)

## Further comments
**Crutch between state systems**
When a legacy model changes it is updated in the redux state as well ([implemented here](https://github.com/nasa-gibs/worldview/blob/redux-toolbar-1576/web/js/main.js#L102-L106)). The syncing between the legacy models and isn't very clean.  I might need to do a sprint and replace all the old models to avoid the cross referencing that is currently being done with the projection model.

**Building new stuff**
You can access all the legacy models using the `state.legacy` object making it very easy to build new things using the redux workflow. [Example](https://github.com/nasa-gibs/worldview/blob/redux-toolbar-1576/web/js/containers/image-download.js#L136)

@nasa-gibs/worldview

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -80,7 +80,6 @@
     "babelify": "^7.3.0",
     "browserstack-capabilities": "^0.4.0",
     "browserstack-local": "^1.3.0",
-    "buster": "github:busterjs/buster#v0.8.0",
```
Comment (ZachTRice) [inline package.json]: 👍

File: package.json [hunk 2]
```diff
@@ -151,14 +154,15 @@
   "dependencies": {
     "@fortawesome/fontawesome-free": "^5.5.0",
     "bluebird": "3.4.6",
-    "bootstrap": "^4.1.0",
+    "bootstrap": "4.1.x",
```
Comment (ZachTRice) [inline package.json]: Curious if there was a strong reason for changing this? Minor version updates are pretty safe with bootstrap as they gracefully deprecate features and show warnings on compile.

File: package.json [hunk 3]
```diff
@@ -170,12 +174,18 @@
     "perfect-scrollbar": "^0.8.1",
     "proj4": "2.3.3",
     "promise-queue": "2.2.3",
-    "react": "^16.3.0",
+    "react": "16.4.x",
```
Comment (ZachTRice) [inline package.json]: 👍

File: package.json [hunk 4]
```diff
@@ -151,14 +154,15 @@
...
-    "bootstrap": "^4.1.0",
+    "bootstrap": "4.1.x",
     "cachai": "^1.0.2",
     "canvg-browser": "^1.0.0",
     "clipboard": "^1.7.1",
     "date-fns": "^1.29.0",
     "file-saver": "^1.3.8",
+    "history": "^4.7.2",
     "icheck": "^1.0.2",
-    "isotope-layout": "^3.0.6",
```
Comment (ZachTRice) [inline package.json]: 👍

File: package.json (hunk 2 above)
Comment (ZachTRice) [reply package.json]: Possibly something to do with `reactstrap`?

File: package.json [hunk 5]
```diff
@@ -186,8 +196,6 @@
     "gifwriter": "./web/ext/gifwriter/gifwriter.js",
     "neuquant": "./web/ext/neuquant/neuquant.js",
     "mobiscroll": "./web/ext/mobiscroll-2.6.0/mobiscroll.js",
-    "isotope-layout": "./node_modules/isotope-layout/dist/isotope.pkgd.js",
-    "isotope": "./node_modules/isotope-layout/dist/isotope.pkgd.js",
```
Comment (ZachTRice) [inline package.json]: 👏

File: web/css/list.css [hunk 6]
```diff
@@ -0,0 +1,23 @@
+.list-group-item {
```
Comment (ZachTRice) [inline web/css/list.css]: Nice work with the CSS changes. I like how list and modal are broken out into separate files; keeps them decoupled like the JS. I'm wondering if we should have a sort of components sub-folder at some point for these type of elements.

File: web/css/notifications.css [hunk 7]
```diff
@@ -4,8 +4,8 @@
   position: relative;
 }
 
-#wv-info-button label::before {
-  content: attr(data-content) '';
```
Comment (ZachTRice) [inline web/css/notifications.css]: Why did the hyphen get removed? `data-*` is the standard HTML attribute syntax.

File: web/js/combine-ui.js [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+import { dataUi } from './data/ui';
+import { animationUi } from './animation/ui';
+import { animationWidget } from './animation/widget';
+import { animationRangeSelect } from './animation/range-select';
+import { animationGif } from './animation/gif';
+// NaturalEvents
+import naturalEventsUI from './natural-events/ui';
+import naturalEventsRequest from './natural-events/request';
+import { debugLayers } from './debug';
+export function combineUi(models, config, MapMouseEvents) {
```
Comment (ZachTRice) [inline web/js/combine-ui.js]: Space between imports and export. Comment the export function

File: web/js/combine-models.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+import { mapModel } from './map/model';
+import { animationModel } from './animation/model';
+import { palettesModel } from './palettes/model';
+import { dataModel } from './data/model';
+import { projectionModel } from './projection/model';
+import { compareModel } from './compare/model';
+import { tourModel } from './tour/model';
+import naturalEventsModel from './natural-events/model';
+
+export function combineModels(config, state) {
```
Comment (ZachTRice) [inline web/js/combine-models.js]: Add comments for export function

File: web/js/combine-middleware.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import thunkMiddleware from 'redux-thunk'; // For ASYNC actions
+import { compact } from 'lodash';
+import { createLogger } from 'redux-logger';
+const loggerMiddleware = createLogger({ collapsed: true });
+export function getMiddleware(isDevelop, locationMiddleware) {
```
Comment (ZachTRice) [inline web/js/combine-middleware.js]: Add space between imports and export. Comment the export function

File: web/js/components/util/image-crop.js [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+// https://stackoverflow.com/a/13139830
+const TRANSPARENT_GIF =
+  'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';
+/*
+ * A react reuseable list component
+ *
+ * @class List
+ * @extends React.Component
+ */
+
```
Comment (ZachTRice) [inline web/js/components/util/image-crop.js]: Remove space

File: web/js/modules/date/util.js [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+    '-' +
+    'T' +
+    date
+      .toISOString()
+      .split('T')[1]
+      .slice(0, -5) +
+    'Z'
+  );
+}
+//   return {
```
Comment (ZachTRice) [inline web/js/modules/date/util.js]: Remove these comments?

File: web/js/modules/image-download/util.js [hunk 13]
```diff
@@ -57,10 +65,10 @@ export function imageUtilCalculateResolution(zoom, isGeoProjection, resolutions)
     }
   }
   return resolution.toString();
-};
+}
 
 /*
- * Retieves avtive layers by day
+ * Retieves active layers by day
```
Comment (ZachTRice) [inline web/js/modules/image-download/util.js]: Don't forget "Retrieves" 😉

File: web/js/modules/image-download/util.js [hunk 14]
```diff
@@ -143,8 +153,42 @@ export function imageUtilGetCoordsFromPixelValues (pixels, map) {
  */
 export function bboxWMS13(lonlats, crs) {
   if (crs === 'EPSG:4326') {
-    return `${lonlats[0][1]},${lonlats[0][0]},${lonlats[1][1]},${lonlats[1][0]}`;
+    return `${lonlats[0][1]},${lonlats[0][0]},${lonlats[1][1]},${
```
Comment (ZachTRice) [inline web/js/modules/image-download/util.js]: Kind of a weird line break here. Prefer how it was before. Not a big deal.

File: web/js/modules/image-download/util.js [hunk 15]
```diff
@@ -143,8 +153,42 @@ export function imageUtilGetCoordsFromPixelValues (pixels, map) {
...
+      lonlats[1][0]
+    }`;
   } else {
-    return `${lonlats[0][0]},${lonlats[0][1]},${lonlats[1][0]},${lonlats[1][1]}`;
+    return `${lonlats[0][0]},${lonlats[0][1]},${lonlats[1][0]},${
```
Comment (ZachTRice) [inline web/js/modules/image-download/util.js]: See comment above

File: web/js/modules/map/util.js [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+  var minLon = 20.6015625 + curHour * (-200.53125 / 23.0);
+  var maxLon = minLon + 159.328125;
+
+  var minLat = -46.546875;
+  var maxLat = 53.015625;
+
+  return [minLon, minLat, maxLon, maxLat];
+}
+
+// export function parseMapExtent(str) {
```
Comment (ZachTRice) [inline web/js/modules/map/util.js]: Clean up unneeded comments in this file

File: web/js/modules/modal/customs.js [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+  TOOLBAR_SNAPSHOT: {
+    headerText: 'Take a Snapshot',
+    backdrop: false,
+    wrapClassName: 'clickable-behind-modal',
+    type: 'selection',
+    offsetRight: '70px',
+    modalClassName: 'toolbar-snapshot-modal toolbar-modal toolbar-medium-modal',
+    bodyComponent: ImageDownload
+  }
+  // NOTIFICATION_LIST_MODAL: {
```
Comment (ZachTRice) [inline web/js/modules/modal/customs.js]: Remove?

File: package.json (hunk 2 above)
Comment (Benjaki2) [reply package.json]: I updated this because I updated it and it broke the tour modal styles. It wasn't too obvious to correct.

**Comment by Benjaki2** [comment]
@ZachTRice I appreciate the detailed review! I will go through it bit by bit. In regards to the `redux-logger`. I have it set up so it will only log in develop. It's really helpful, it gives you the before and after state of each action, but that could be annoying so maybe I should make it so it only logs if you have it configured to log, somehow?

**Comment by ZachTRice** [comment]
AS long as it is only showing in develop, then I think it is fine. It is a lot of logging so if there was a way to trim it back that might be good, but not a huge concern. Worth taking a look at but don't spend a bunch of time implementing a work around.

File: web/js/modules/image-download/util.js (hunk 15 above)
Comment (Benjaki2) [reply web/js/modules/image-download/util.js]: That is weird. I'm thinking `prettier.io` must have done that one

File: web/js/modules/modal/customs.js (hunk 17 above)
Comment (Benjaki2) [reply web/js/modules/modal/customs.js]: Good call. I need to remove this whole file.

File: web/css/list.css (hunk 6 above)
Comment (Benjaki2) [reply web/css/list.css]: I like it.

File: web/css/notifications.css (hunk 7 above)
Comment (Benjaki2) [reply web/css/notifications.css]: I didn't think `react` hyphens with attributes. I was wrong. Good call 👍

**Comment by Benjaki2** [comment]
@ZachTRice It will probably get in the way of a lot of performance profiling and other things. I could comment it out and people can enable it to use it when they need it.

**Comment by Benjaki2** [comment]
@ZachTRice Actually i put in a `npm run watch:log`

**Comment by Benjaki2** [comment]
@edplato Thanks for the in depth review! I believe I have addressed those issues now so please take another look when you get a chance.

**Pull request merged**